### PR TITLE
sql/catalog: address session based leasing bugs

### LINKED ISSUE #120445: pkg/sql/logictest/tests/local/local_test: TestLogic_schema_repair failed
pkg/sql/logictest/tests/local/local_test.TestLogic_schema_repair [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/14383219?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/14383219?buildTab=artifacts#/) on master @ [ee5768958fcf7940c73beae06d55377b1b3b3ed8](https://github.com/cockroachdb/cockroach/commits/ee5768958fcf7940c73beae06d55377b1b3b3ed8):

```
=== RUN   TestLogic_schema_repair
    test_log_scope.go:170: test logs captured to: /artifacts/tmp/_tmp/ab11c9cd60408977be9520945b6e4e92/logTestLogic_schema_repair2224741283
    test_log_scope.go:81: use -show-logs to present logs inline
    logic.go:4308: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/ab11c9cd60408977be9520945b6e4e92/logTestLogic_schema_repair2224741283
--- FAIL: TestLogic_schema_repair (1.69s)
=== RUN   TestLogic_schema_repair/force_delete_data
    logic.go:3011: let $t_id = 108
    logic.go:3011: let $parentID = 104
    logic.go:3011: let $parentSchemaID = 105
    logic.go:3011: let $json = {"table": {"columns": [{"id": 1, "name": "v", "nullable": true, "type": {"family": "IntFamily", "oid": 20, "width": 64}}, {"defaultExpr": "unique_rowid()", "hidden": true, "id": 2, "name": "rowid", "type": {"family": "IntFamily", "oid": 20, "width": 64}}], "createAsOfTime": {}, "families": [{"columnIds": [1, 2], "columnNames": ["v", "rowid"], "defaultColumnId": 1, "name": "primary"}], "formatVersion": 3, "id": 108, "modificationTime": {}, "name": "forcedeletemydata", "nextColumnId": 3, "nextConstraintId": 2, "nextFamilyId": 1, "nextIndexId": 2, "nextMutationId": 1, "parentId": 104, "primaryIndex": {"constraintId": 1, "createdAtNanos": "1710368410157421000", "encodingType": 1, "foreignKey": {}, "geoConfig": {}, "id": 1, "interleave": {}, "keyColumnDirections": ["ASC"], "keyColumnIds": [2], "keyColumnNames": ["rowid"], "name": "forcedeletemydata_pkey", "partitioning": {}, "sharded": {}, "storeColumnIds": [1], "storeColumnNames": ["v"], "unique": true, "version": 4}, "privileges": {"ownerProto": "root", "users": [{"privileges": "2", "userProto": "admin", "withGrantOption": "2"}, {"privileges": "2", "userProto": "root", "withGrantOption": "2"}], "version": 3}, "replacementOf": {"time": {}}, "unexposedParentSchemaId": 105, "version": "1"}}
    logic.go:2965: 
         
        /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/1232/execroot/com_github_cockroachdb_cockroach/bazel-out/aarch64-fastbuild/bin/pkg/sql/logictest/tests/local/local_test_/local_test.runfiles/com_github_cockroachdb_cockroach/pkg/sql/logictest/testdata/logic_test/schema_repair:190: SELECT * FROM forcedeletemydata ORDER BY v ASC
        expected success, but found
        (XXUUU) unexpected value: raw_bytes:"\257\237s\312\n3\002" timestamp:<wall_time:1710368410165452178 > 
    logic.go:2207: 
         /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/1232/execroot/com_github_cockroachdb_cockroach/bazel-out/aarch64-fastbuild/bin/pkg/sql/logictest/tests/local/local_test_/local_test.runfiles/com_github_cockroachdb_cockroach/pkg/sql/logictest/testdata/logic_test/schema_repair:194: too many errors encountered, skipping the rest of the input
[22:20:10] --- done: /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/1232/execroot/com_github_cockroachdb_cockroach/bazel-out/aarch64-fastbuild/bin/pkg/sql/logictest/tests/local/local_test_/local_test.runfiles/com_github_cockroachdb_cockroach/pkg/sql/logictest/testdata/logic_test/schema_repair with config local: 26 tests, 2 failures
--- total: 26 tests, 2 failures
    --- FAIL: TestLogic_schema_repair/force_delete_data (0.11s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/sql-queries
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestLogic_schema_repair.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

Jira issue: CRDB-36691

### LINKED ISSUE #120675: upgrade/upgrades: TestSessionBasedLeaseUpgrade failed
upgrade/upgrades.TestSessionBasedLeaseUpgrade [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/14449002?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/14449002?buildTab=artifacts#/) on master @ [91c1d5cbb5a56cd9c082d78814cd03c5471dddfa](https://github.com/cockroachdb/cockroach/commits/91c1d5cbb5a56cd9c082d78814cd03c5471dddfa):

```
=== RUN   TestSessionBasedLeaseUpgrade
    test_log_scope.go:170: test logs captured to: /artifacts/tmp/_tmp/19fe8462015815b7be025b680724a827/logTestSessionBasedLeaseUpgrade3482969845
    test_log_scope.go:81: use -show-logs to present logs inline
    test_server_shim.go:144: cluster virtualization disabled in global scope due to issue: #76378 (expected label: C-bug)
    v24_1_session_based_lease_test.go:104: 
        	Error Trace:	github.com/cockroachdb/cockroach/pkg/upgrade/upgrades_test/pkg/upgrade/upgrades/v24_1_session_based_lease_test.go:104
        	Error:      	Received unexpected error:
        	            	pq: transaction committed but schema change aborted with error: (42703): count-leases: column "desc_id" does not exist
        	Test:       	TestSessionBasedLeaseUpgrade
    panic.go:626: -- test log scope end --
--- FAIL: TestSessionBasedLeaseUpgrade (2.66s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/release-eng
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestSessionBasedLeaseUpgrade.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)
</sub>

Jira issue: CRDB-36815

### LINKED ISSUE #120735: roachtest: schemachange/mixed-versions failed [leases.desc_id column does not exist; issue with leasing migration]
roachtest.schemachange/mixed-versions [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RoachtestNightlyGceBazel/14467090?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RoachtestNightlyGceBazel/14467090?buildTab=artifacts#/schemachange/mixed-versions) on master @ [f0116ea373a2b87155e7f0264df4f783ce177360](https://github.com/cockroachdb/cockroach/commits/f0116ea373a2b87155e7f0264df4f783ce177360):

```
(mixedversion.go:592).Run: mixed-version test failure while running step 12 (run "run schemachange workload and validation in mixed version"): full command output in run_060528.985987797_n4_COCKROACHRANDOMSEED5.log: COMMAND_PROBLEM: exit status 1
test artifacts and logs in: /artifacts/schemachange/mixed-versions/run_1
```

Parameters:
 - <code>ROACHTEST_arch=amd64</code>
 - <code>ROACHTEST_cloud=gce</code>
 - <code>ROACHTEST_coverageBuild=false</code>
 - <code>ROACHTEST_cpu=4</code>
 - <code>ROACHTEST_encrypted=false</code>
 - <code>ROACHTEST_metamorphicBuild=false</code>
 - <code>ROACHTEST_ssd=0</code>
<details><summary>Help</summary>
<p>

See: [roachtest README](https://github.com/cockroachdb/cockroach/blob/master/pkg/cmd/roachtest/README.md)

See: [How To Investigate \(internal\)](https://cockroachlabs.atlassian.net/l/c/SSSBr8c7)

See: [Grafana](https://go.crdb.dev/roachtest-grafana/teamcity-14467090/schemachange-mixed-versions/1710914551716/1710914873523)

</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #119904 roachtest: schemachange/mixed-versions failed [job-rows-batch-insert error; let&#39;s ignore that error] [C-test-failure O-roachtest O-robot P-3 T-sql-foundations branch-release-23.1.17-rc]
</p>
</details>
/cc @cockroachdb/sql-foundations
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*schemachange/mixed-versions.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)
</sub>

Jira issue: CRDB-36844

### PR DESCRIPTION
**Pull request by fqazi**
This patch addresses the following issues:
1. Previously the session based leasing upgrade had a race condition against the leasing count queries, which could cause them to incorrect expect the wrong version of the descriptor. To address this, we are going to start using the system database descriptor version to determine if synthetic descriptors should be used (instead of cluster settings).
2. The repair_schema logictest had an intermittent failure because it was force deleting and injecting descriptors instantly. This runs into an existing design limitation of the lease manager where deleted descriptors are instantly cleaned up from memory, but asynchronously removed from storage. When we inject the descriptor with the exact same versions, the lease manager would detect a duplicate row and fails renewing (in the session based model). This condition can only be hit by repair queries and in the expiry based model would lead to duplicate rows. For the purpose of testing we are going to intentionally give a new version when adding the descriptor back, since the sequence in general dangerous with repair queries (i.e. e need to confirm that the lease is no longer in use before injecting the descriptor again or set a new version).

Fixes: #120445 
Fixes #120735 
Fixes #120675

### CODE REVIEW

**Review by rafiss — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/120780)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @fqazi)

___
*[`pkg/sql/catalog/lease/count.go` line 46 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/120780#-NtUNX6vCZjVAaYl9UCl:-NtUNX6vCZjVAaYl9UCm:bdn1w77) ([raw file](https://github.com/cockroachdb/cockroach/blob/792a91be446af93bd487388e5898c4a74d18e957/pkg/sql/catalog/lease/count.go#L46)):*
> ```Go
> ) (int, error) {
> 	// Indicates if the leasing descriptor has been upgraded for session based
> 	// leasing. Note: Unit tests will never provide this object, so assume
> ```

nit: what is "this object" referring to?
___
*[`pkg/sql/catalog/lease/count.go` line 54 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/120780#-NtUNkdX5h0gFSlN07tN:-NtUNkdX5h0gFSlN07tO:b-69z5c6) ([raw file](https://github.com/cockroachdb/cockroach/blob/792a91be446af93bd487388e5898c4a74d18e957/pkg/sql/catalog/lease/count.go#L54)):*
> ```Go
> 		v := settings.Version.ActiveVersion(ctx).Version
> 		leasingDescIsSessionBased = hasSessionBasedLeasingDesc(&v)
> 	}
> ```

nit: more concise as
```
var systemDBVersion *roachpb.Version
if cachedDatabaseRegions != nil {
  systemDBVersion = cachedDatabaseRegions.GetDatabaseVersion()
} else {
  systemDBVersion =  &settings.Version.ActiveVersion(ctx).Version
}
leasingDescIsSessionBased := systemDBVersion != nil && 
  systemDBVersion.AtLeast(clusterversion.V24_1_SessionBasedLeasingUpgradeDescriptor.Version())
```
(and no need for the separate hasSessionBasedLeasingDesc function)
___
*[`pkg/sql/catalog/lease/count.go` line 76 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/120780#-NtUNHoU6Y8H645yrS6n:-NtUNHoV9UYPN_BxrIN1:b-fdbztw) ([raw file](https://github.com/cockroachdb/cockroach/blob/792a91be446af93bd487388e5898c4a74d18e957/pkg/sql/catalog/lease/count.go#L76)):*
> ```Go
> 		// The leasing descriptor is not session based yet, so we need to inject
> 		// in synthetically.
> 		if leasingDescIsSessionBased {
> ```

why did this condition get flipped? it looks to me like keeping the `!` is correct
___
*[`pkg/sql/regions/cached_db_regions.go` line 72 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/120780#-NtUOpa93vRGdH8tu6f4:-NtUOpa93vRGdH8tu6f5:b-hsx712) ([raw file](https://github.com/cockroachdb/cockroach/blob/792a91be446af93bd487388e5898c4a74d18e957/pkg/sql/regions/cached_db_regions.go#L72)):*
> ```Go
> 
> // GetDatabaseVersion helper to get the system database version.
> func (c *CachedDatabaseRegions) GetDatabaseVersion() *roachpb.Version {
> ```

nit: `GetSystemDatabaseVersion` could be a more clear name
___
*[`pkg/sql/logictest/testdata/logic_test/schema_repair` line 198 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/120780#-NtUMnck726csylpO6Ww:-NtUMnclAORslMzHcgUb:b2wps1x) ([raw file](https://github.com/cockroachdb/cockroach/blob/8e3ee722a04ff0e60db7249527ec7ec5c6414fa0/pkg/sql/logictest/testdata/logic_test/schema_repair#L198)):*
> ```
> FROM
> 	ROWS FROM (
> 		crdb_internal.unsafe_upsert_descriptor(
> ```

would it work to modify unsafe_upsert_descriptor to always bump the version?

**Review by fqazi — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/120780)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @rafiss)

___
*[`pkg/sql/catalog/lease/count.go` line 46 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/120780#-NtUNX6vCZjVAaYl9UCl:-NtWVI-B-rFHfj5XLppb:bfbky73) ([raw file](https://github.com/cockroachdb/cockroach/blob/792a91be446af93bd487388e5898c4a74d18e957/pkg/sql/catalog/lease/count.go#L46)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

nit: what is "this object" referring to?
</blockquote></details>

Updated this to:

// Indicates if the leasing descriptor has been upgraded for session based  
// leasing. Note: Unit tests will never provide cached database regions  
// so resolve the version from the cluster settings.
___
*[`pkg/sql/catalog/lease/count.go` line 76 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/120780#-NtUNHoU6Y8H645yrS6n:-NtWVq-FAQQuSKk0LhtR:bicxkp8) ([raw file](https://github.com/cockroachdb/cockroach/blob/792a91be446af93bd487388e5898c4a74d18e957/pkg/sql/catalog/lease/count.go#L76)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

why did this condition get flipped? it looks to me like keeping the `!` is correct
</blockquote></details>

So, the  systemschema.LeaseTable\_V23\_2() descriptor is the expiry based one. It should only be injected **if** the descriptor is session based. Otherwise, no synthetic descriptor should be needed.

The comment above is also wrong, so I fixed it:
// The leasing descriptor is session based, so we need to inject  
// expiry based descriptor synthetically.
___
*[`pkg/sql/logictest/testdata/logic_test/schema_repair` line 198 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/120780#-NtUMnck726csylpO6Ww:-NtWWQSrCCB6g-UwtDhT:b3jn7p2) ([raw file](https://github.com/cockroachdb/cockroach/blob/8e3ee722a04ff0e60db7249527ec7ec5c6414fa0/pkg/sql/logictest/testdata/logic_test/schema_repair#L198)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

would it work to modify unsafe_upsert_descriptor to always bump the version?
</blockquote></details>

Yeah, that would also work as an alternative. Normally our run book scenarios have us poking at the version number manually and it would remove a step too. Maybe the logic could be:

if originalVersion == updateVersion {
updateVersion = updateVersion +1
}

**Review by rafiss — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/120780)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @fqazi)

___
*[`pkg/sql/logictest/testdata/logic_test/schema_repair` line 198 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/120780#-NtUMnck726csylpO6Ww:-NtWbi8w5pYAW4UbJqh0:b-c7n8yo) ([raw file](https://github.com/cockroachdb/cockroach/blob/8e3ee722a04ff0e60db7249527ec7ec5c6414fa0/pkg/sql/logictest/testdata/logic_test/schema_repair#L198)):*
<details><summary><i>Previously, fqazi (Faizan Qazi) wrote…</i></summary><blockquote>

Yeah, that would also work as an alternative. Normally our run book scenarios have us poking at the version number manually and it would remove a step too. Maybe the logic could be:

if originalVersion == updateVersion {
updateVersion = updateVersion +1
}
</blockquote></details>

yeah, let's try that so we don't need this special test version bump

File: pkg/sql/repair.go [hunk 1]
```diff
@@ -174,6 +176,22 @@ func (p *planner) UnsafeUpsertDescriptor(
 		p.Descriptors().SkipValidationOnWrite()
 	}
 
+	// If we are pushing out a brand new descriptor confirm that no leases
+	// exist before we publish it. This could happen if we did an unsafe delete,
+	// since we will not wait for all leases to expire. So, as a safety force the
```
Comment (rafiss) [inline pkg/sql/repair.go:181]: nit: "as a safety force the for unsafe upserts" might be missing a word

**Review by fqazi — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/120780)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @rafiss)

___
*[`pkg/sql/repair.go` line 181 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/120780#gh-1534466747:-NtXL3ucEKB8J1rL8zeW:b-y8md2h) ([raw file](https://github.com/cockroachdb/cockroach/blob/24c1353deaadc94e92e6fc8b150cd0c0f6a5aff6/pkg/sql/repair.go#L181)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

nit: "as a safety force the for unsafe upserts" might be missing a word
</blockquote></details>

Should have been:

// If we are pushing out a brand new descriptor confirm that no leases  
// exist before we publish it. This could happen if we did an unsafe delete,  
// since we will not wait for all leases to expire. So, as a safety force the  
// unsafe upserts to wait for no leases to exist on this descriptor.
___
*[`pkg/sql/logictest/testdata/logic_test/schema_repair` line 198 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/120780#-NtUMnck726csylpO6Ww:-NtX44XEDUVHUInRl71Y:bgfu84a) ([raw file](https://github.com/cockroachdb/cockroach/blob/8e3ee722a04ff0e60db7249527ec7ec5c6414fa0/pkg/sql/logictest/testdata/logic_test/schema_repair#L198)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

yeah, let's try that so we don't need this special test version bump
</blockquote></details>

We already have a check with the version bump. I'm going to add safety to instead have the non-force mode to wait for old versions to expire on the unsafe upsert (this code should only be hit in the scenario where a descriptor is being added back).

**Comment by fqazi** [comment]
@rafiss TFTR!

bors r+

**Pull request merged**